external/fv3fit/fv3fit/pytorch/cyclegan/modules.py

```diff
@@ -144,6 +143,199 @@ def single_tile_convolution(
     return FoldTileDimension(conv)
 
 
+def halo_convolution(
+    in_channels: int,
+    out_channels: int,
+    kernel_size: int,
+    output_padding: int = 0,
+    stride: int = 1,
+    stride_type: Literal["regular", "transpose"] = "regular",
+    bias: bool = True,
+) -> ConvolutionFactory:
+    """
+    Construct a convolutional layer that pads with halo data before applying conv2d.
+
+    Padding is such that the output shape should be the input shape divided by the
+    stride for regular convolution. For transpose convolution, the output shape
+    will be the input shape multiplied by the stride.
+
+    Layer takes in and returns tensors of shape [batch, tile, channels, x, y].
+
+    Args:
+        kernel_size: size of the convolution kernel
+        output_padding: argument used for transpose convolution
+        stride: stride of the convolution
+        stride_type: type of stride, one of "regular" or "transpose"
+        bias: whether to include a bias vector in the produced layers
+    """
+    if stride_type == "transpose":
+        padding = int((kernel_size - 1) // 2 * stride)
+    else:
+        padding = int((kernel_size - 1) // 2)
+    append = AppendHalos(n_halo=padding)
+    conv = single_tile_convolution(
+        in_channels=in_channels,
+        out_channels=out_channels,
+        kernel_size=kernel_size,
+        output_padding=output_padding,
+        stride=stride,
+        stride_type=stride_type,
+        bias=bias,
+    )
+    if stride_type == "transpose":
+        # have to crop halo points from the output, as pytorch has no option to
+        # only output a subset of the domain for ConvTranspose2d
+        conv = nn.Sequential(
+            conv, Crop(n_halo=padding * stride + int(kernel_size - 1) // 2)
+        )
+    return nn.Sequential(append, conv)
+
+
+def cpu_only(method):
+    """
+    Decorator to mark a method as only being supported on the CPU.
+    """
+
+    @functools.wraps(method)
+    def wrapper(self, *args, **kwargs):
+        original_device = args[0].device
+        args = [arg.cpu() if isinstance(arg, torch.Tensor) else arg for arg in args]
+        kwargs = {
+            k: v.cpu() if isinstance(v, torch.Tensor) else v for k, v in kwargs.items()
+        }
+        return method(self, *args, **kwargs).to(original_device)
+
+    return wrapper
+
+
+class Crop(nn.Module):
+    def __init__(self, n_halo):
+        super(Crop, self).__init__()
+        self.n_halo = n_halo
+
+    @cpu_only
+    def forward(self, x):
+        return x[
+            ..., self.n_halo : -self.n_halo, self.n_halo : -self.n_halo
+        ].contiguous()
+
+
+class AppendHalos(nn.Module):
```

Comment:
I'm glad this passes the test against the other implementation that uses the CubedSphereCommunicator, otherwise this would be a bit of a scary thing to debug.